toolkit/tools/scheduler/schedulerutils/buildlist.go

```diff
@@ -21,84 +23,73 @@ import (
+// If the string is NOT a spec name, it will check, if the string is a package present in the graph.
+// If the package is not present in the graph, it will return an error.
+// All packages without a build node are considered invalid.
+//
+// Note: since "SRPM_PACK_LIST" can work only with spec names, spec names take priority over package names
+// so that passing "X" to "SRPM_PACK_LIST", "PACKAGE_IGNORE_LIST", and "*_BUILD_LIST" arguments targets the same set of packages.
+func PackageNamesToBuiltPackages(packageOrSpecNames []string, dependencyGraph *pkggraph.PkgGraph) (packageVers []*pkgjson.PackageVer, err error) {
+	logger.Log.Debugf("Converting package/spec names to build nodes' PackageVers: %v", packageOrSpecNames)
+
+	packageVersMap := make(map[*pkgjson.PackageVer]bool)
+	specToPackageNodes := make(map[string][]*pkggraph.PkgNode)
+
+	for _, node := range dependencyGraph.AllBuildNodes() {
+		specToPackageNodes[node.SpecName()] = append(specToPackageNodes[node.SpecName()], node)
 	}
 
-	packageList, err = installutils.PackageNamesFromConfig(cfg)
-	if err != nil {
-		return
-	}
-
-	// Add kernel packages from KernelOptions
-	packageList = append(packageList, installutils.KernelPackages(cfg)...)
-
-	return
-}
-
-// filterLocalPackagesOnly returns the subset of packageVersionsInConfig that only contains local packages.
-func filterLocalPackagesOnly(packageVersionsInConfig []*pkgjson.PackageVer, inputGraph string) (filteredPackages []*pkgjson.PackageVer, err error) {
-	logger.Log.Debug("Filtering out external packages from list of packages extracted from the image config file.")
-
-	dependencyGraph := pkggraph.NewPkgGraph()
-	err = pkggraph.ReadDOTGraphFile(dependencyGraph, inputGraph)
-	if err != nil {
-		return
-	}
-
-	for _, pkgVer := range packageVersionsInConfig {
-		pkgNode, _ := dependencyGraph.FindBestPkgNode(pkgVer)
-
-		// A pkgNode for a local package has the following characteristics:
-		// 1) The pkgNode exists in the graph (is not nil).
-		// 2) The pkgNode doesn't have the 'StateUnresolved' or 'StateCached' state. These are reserved for external dependencies nodes.
-		if pkgNode != nil && pkgNode.RunNode.State != pkggraph.StateUnresolved && pkgNode.RunNode.State != pkggraph.StateCached {
-			filteredPackages = append(filteredPackages, pkgVer)
+	for _, packageOrSpecName := range packageOrSpecNames {
+		if nodes, ok := specToPackageNodes[packageOrSpecName]; ok {
+			logger.Log.Debugf("Name '%s' matched a spec name. Adding all packages built from this spec to the list.", packageOrSpecName)
+			for _, pkg := range nodes {
+				packageVersMap[pkg.VersionedPkg] = true
+			}
 		} else {
-			logger.Log.Debugf("Found external package to filter out: %v.", pkgVer)
+			foundNode, err := dependencyGraph.FindBestPkgNode(&pkgjson.PackageVer{Name: packageOrSpecName})
+			if err != nil {
+				logger.Log.Errorf("Failed while searching the dependency graph for package '%s', error: %s", packageOrSpecName, err)
+				return nil, err
+			}
```

Comment:
I've been thinking we should use wrapped errors more, there are several spots in the code were we completely lose our error state and overwrite it. (We also can compare against the wrapped errors, although we don't really do that much in our code since generally any error is effectively fatal) Thoughts?
```suggestion
			if err != nil {
				err = fmt.Errorf("failed while searching the dependency graph for package '%s': %w", packageOrSpecName, err)
				return nil, err
			}
```

Reply:
In general I don't really mind the way we print errors in most places right now - the "stacked" approach. It is similar to printing a call stack and we don't run into a risk of printing one, insanely long error message, which can make it hard to read. 

As far as I remember, we intentionally don't try to parse the errors and then decide how to treat them. This would have the effect of producing exception-/error-driven code (not to be confused with exception-driven development), which doesn't help with code clarity. In general, I'm strongly enforcing having as few ignorable errors as possible, since this way the developer doesn't have to consider the possibility if a returned error is something they can tolerate or not. Instead, it is a clear message that something went wrong and we have to bail.

Reply:
Agreed about not trying to parse the errors if at all possible. I've just noticed a lot of places where we do
```
if err != nil {
    // Forgot to write a logger.Log.Errorf()
    return err
}
```
Or some variation. The problem I have with using the print statements is if we aren't super verbose we can lose track of where the caller is. Not saying %w is a magic bullet but I like it more so far.

Reply:
OK, now I think I better understand what you had in mind when you mentioned the unwrapped errors. I think I also like the idea of wrapping error, let me see how this may looks like here.